quinn/src/connection.rs

```diff
@@ -933,40 +942,66 @@ pub(crate) struct State {
     /// Number of live handles that can be used to initiate or handle I/O; excludes the driver
     ref_count: usize,
     socket: Arc<dyn AsyncUdpSocket>,
+    io_poller: Pin<Box<dyn UdpPoller>>,
     runtime: Arc<dyn Runtime>,
+    /// We buffer a transmit when the underlying I/O would block
+    buffered_transmit: Option<udp::Transmit>,
 }
 
 impl State {
-    fn drive_transmit(&mut self) -> bool {
+    fn drive_transmit(&mut self, cx: &mut Context) -> io::Result<bool> {
         let now = Instant::now();
         let mut transmits = 0;
 
         let max_datagrams = self.socket.max_transmit_segments();
         let capacity = self.inner.current_mtu();
         let mut buffer = BytesMut::with_capacity(capacity as usize);
 
-        while let Some(t) = self.inner.poll_transmit(now, max_datagrams, &mut buffer) {
-            transmits += match t.segment_size {
-                None => 1,
-                Some(s) => (t.size + s - 1) / s, // round up
+        loop {
+            // Retry the last transmit, or get a new one.
+            let t = match self.buffered_transmit.take() {
+                Some(t) => t,
+                None => match self.inner.poll_transmit(now, max_datagrams, &mut buffer) {
+                    Some(t) => {
+                        transmits += match t.segment_size {
+                            None => 1,
+                            Some(s) => (t.size + s - 1) / s, // round up
+                        };
+
+                        let buffer = buffer.split_to(t.size).freeze();
+                        udp_transmit(t, buffer)
+                    }
+                    None => break,
+                },
             };
-            // If the endpoint driver is gone, noop.
-            let size = t.size;
-            let _ = self.endpoint_events.send((
-                self.handle,
-                EndpointEvent::Transmit(t, buffer.split_to(size).freeze()),
-            ));
+
+            if self.io_poller.as_mut().poll_writable(cx)?.is_pending() {
+                // Retry after a future wakeup
+                self.buffered_transmit = Some(t);
+                return Ok(false);
+            }
+
+            let result = self.socket.try_send(std::slice::from_ref(&t));
+            if result
+                .as_ref()
+                .map_or_else(|e| e.kind() == io::ErrorKind::WouldBlock, |&n| n == 0)
+            {
+                // Retry immediately
+                self.buffered_transmit = Some(t);
+                continue;
+            }
+            result?;
```

Comment:
This is pretty dense. This is one line more but maybe easier to follow?

```rust
let retry = match self.socket.try_send(std::slice::from_ref(&t)) {
    Ok(0) => true,
    Err(e) if e.kind() == io::ErrorKind::WouldBlock => true,
    Err(e) => return Err(err),
};

if retry {
    self.buffered_transmit = Some(t);
    continue;
}
```

Reply:
Oh, that's much better indeed.